web/src/shared/components/pedigree/TangledTree.tsx

```diff
@@ -11,19 +15,672 @@ interface PedigreeEntry {
+            }
+            // try add person above children
+            const levels = keyedData[next].children
+                .map((n) => findInHeirarchy(n, toReturn))
+                .filter((a) => a > -1)
+            if (levels.length) {
+                toReturn[min(levels) - 1] = [
+                    ...toReturn[min(levels) - 1],
+                    {
+                        id: next,
+                        ...((keyedData[next].paternal_id || keyedData[next].maternal_id) && {
+                            parentsList: [
+                                keyedData[next].paternal_id,
+                                keyedData[next].maternal_id,
+                            ].filter((l) => l),
+                        }),
+                    },
+                ]
+                yetToSee.delete(next)
+                updatedList = true
+                break
+            }
+            // try add child below parents
+            const parentLevel = max([
+                findInHeirarchy(keyedData[next].maternal_id, toReturn),
+                findInHeirarchy(keyedData[next].paternal_id, toReturn),
+            ])
+            if (parentLevel > -1) {
+                toReturn[parentLevel + 1] = [
+                    ...toReturn[parentLevel + 1],
+                    {
+                        id: next,
+                        ...((keyedData[next].paternal_id || keyedData[next].maternal_id) && {
+                            parentsList: [
+                                keyedData[next].paternal_id,
+                                keyedData[next].maternal_id,
+                            ].filter((m) => m),
+                        }),
+                    },
+                ]
+                yetToSee.delete(next)
+                updatedList = true
+                break
+            }
+        }
+        if (!updatedList) {
+            break
+        }
+        updatedList = false
+    }
+
+    const nodeDefaults = {
+        x: 0,
+        y: 0,
+        parents: [],
+        level: -1,
+        bundles: [],
+        links: [],
+        parentsList: [],
+    }
+
+    return toReturn.map((l) => ({
+        nodes: l.map((m) => ({ ...nodeDefaults, ...m })),
+        bundles: [],
+    }))
+}
+
 interface RenderPedigreeProps {
     data: PedigreeEntry[]
     click?(e: string): void
 }
 
-const TangledTree: React.FunctionComponent<RenderPedigreeProps> = ({ data, click }) => (
-    <>
-        Temporary Pedigree component
-        <br />
-        {JSON.stringify(data)}
-        <br />
-        {JSON.stringify(click)}
-    </>
-)
+const TangledTree: React.FunctionComponent<RenderPedigreeProps> = ({ data, click }) => {
+    const [tree, setTree] = React.useState<NodeList[]>([])
+    const [keyedData, setKeyedData] = React.useState<{
+        [name: string]: PedigreeEntry
+    }>()
+    const onClick =
+        click ??
+        function defaultClick() {
+            return undefined
+        }
+
+    React.useEffect(() => {
```

Comment:
You don't need a `useEffect` or `useState` hooks. Try using derived properties instead, which will be re-computed any time that the props change (like you have done with `onClick`). This will clean up your code considerably.

```es6
const tree: NodeList[] = data ? formatData(data) : []
const keyedData: {[name: string]: PedigreeEntry} = data ? data.reduce(..., {}) : {}
```